### ISSUE TITLE
HelpDefs are not generated when defs of different classes have identical names.

### ISSUE THREAD

**Issue by Karel-Kroeze** [RimWorldCCLTeam/CommunityCoreLibrary#38]
For example, the Hospital Bed has a building (which is resolved and created first), and an identically named research def.

The first will create a helpdef, the second will not - the research def will simply not generate a helpdef.

The offending bit of code is;

```
var helpDef = helpDefs.Find( h => (
    (.keyDef == thingDef.defName )
) );
```

which is called in each and every helpdef resolver, and in essence only checks by the 'generating' defs name, not it's type.

**Comment by ForsakenShell** [comment]
I didn't even think of different classes with the same defname, simple enough to resolve, however.  Will be sure to do a fix soon.

**Comment by Karel-Kroeze** [comment]
I've got a local fix by setting keyDef to actually be the generating def. Still finetuning the current version of the research info embedded in CCL, will do a PR with a fix and a tweaked research info tab tonight or tomorrow.

I plan on overhauling the way the descriptions are generated, instead of creating a large string in the resolvers, I'ld like to attach a list of defs to the various unlocks/requires/etc segments. Description strings can then be built on-the-fly, but we can also cross-link based on the defs. Still thinking on how to implement that properly.

**Comment by ForsakenShell** [comment]
The reason I did it by building the strings is that (a) the strings are relatively small given the limited data set and (b) it's done once so it save processing during game play.  If you have it fixed at your end I'll leave it at that then.

**Comment by Karel-Kroeze** [comment]
I may very well be wrong, but would storing a list of references to 'related' helpdefs be much more taxing on memory than sorting strings? 

As to processing during gameplay - both the research and help screens pause the game, I doubt it's going to have a massive impact...

I still need to think on how to implement this, right now I'm thinking of adding a List<pair<role, object>> for the unlocks/requires/etc sections, where role is the specific section object should be listed in, and object can either be a helpdef, or in cases where this does not exist (e.g. planting tags) a string. Another issue is that this requires either doing a double pass of the helpdef resolving (first pass to create all the defs, second to link them), or creating helpdefs as the link is created. Considering this is a one-time op, I may just take the easy way out and do a double pass.

**Comment by ForsakenShell** [comment]
Resolved in #41
